designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,326 @@
+- Repo: https://github.com/eslint/eslint
+- Start Date: 2025-07-10
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: ST-DDT
+
+# Rule Tester Assertions Options
+
+## Summary
+
+Add options that control which assertions are required for each `RuleTester` test case.
+
+## Motivation
+
+In most eslint(-plugin)s' rules the error assertions are different from each other.
+Adding options that could be set/shared when configuring the rule tester would ensure a common base level of assertions is met throughout the (plugin's) project.
+The options could be defined on two levels. On `RuleTester`'s `constructor` effectively impacting all tests, or on the test `run` method itself, so only that set is affected.
+
+## Detailed Design
+
+### Variant 1 - Constructor based options
+
+````ts
+new RuleTester(testerConfig: {...}, assertionOptions: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     * 
+     * @default false
+     */
+    requireMessage: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+     * 
+     * @default false
+     */
+    requireLocation: boolean;
+} = {});
+````
+
+### Variant 2 - Test method based options
+
+````ts
+ruleTester.run("rule-name", rule, tests, assertionOptions: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+     * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+     * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+     * 
+     * @default false
+     */
+    requireMessage: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
+     * 
+     * @default false
+     */
+    requireLocation: boolean;
+});
+````
```

Comment:
I'm slightly more in favor of this variant (Variant 2 - Test method based options), as it allows for multiple runs with different assertion options. As a variant of this variant, maybe we could consider adding `assertionOptions` property to the `tests` object, instead of it being the last parameter of the method call, so that it doesn't need to appear at the very end of the test run, after all tests to which it applies.

```js
ruleTester.run("rule-name", rule, {
    assertionOptions: { /* ... */ },
    valid: [ /* ... */ ],
    invalid: [ /* ... */ ],
});
```

Reply:
Done